onnx/onnx.proto

```diff
@@ -212,6 +243,165 @@ message TensorProto {
+  optional int64 ir_version = 1;
+
+  // The optional version of the framework runtime that generates this graph.
+  // This producer_version has the same format as ir_version. 
+  optional int64 producer_version = 2;
+
+  // The optional name of the framework used to generate this graph in the form
+  // "framework_name[-tag]". Tag is optional and provides additional
+  // information such as `alpha` or `beta` or `rc3`.
+  optional string producer_tag = 3;
+
+  // An optional version identifier used to track evolution of this library.
+  // This model_version has the same format as ir_version. 
+  optional int64 model_version = 4;
+
+  // The optional name of the author who created the library.
+  optional string model_author = 5;
+
+  // Optional licensing information concerning use or origination of the library.
+  optional string model_license = 6;
+
+  // The name of the library.
+  optional string name = 7;   // namespace Library
+
+  // Domain of the library.
+  // We use reverse domain names as name space indicators. For example:
+  // `com.facebook.fair` or `com.microsoft.cognitiveservices`
+  //
+  // Together with `name` and `model_version`, this forms the unique identity of
+  // the library.
+  optional string domain = 8;
+
+  // An optional human-readable documentation for this graph.
+  // This string MAY contain markdown syntax in GFM form.
+  optional string doc_string = 9;
+
+  // The operators declared by this library. 
+  repeated OperatorDeclProto operator = 10;
+
+  // A given name may appear at most once in operator.
+  // When refering to an operator from outside of this library, the op_type
+  // field must equal:
+  //      LibraryProto.domain + "." + LibraryProto.name + "." OperatorDeclProto.name  
+  // or
+  //      LibraryProto.domain + "." + LibraryProto.name + "." FunctionDefProto.name  
+
+  // Imported libraries are referenced as a collection of strings in the form of absolute
+  // URIs or relative paths. Where such relative paths are rooted is defined by tools and
+  // runtime implementations.
+  repeated string imported_libraries = 12;  
+};
+
+
+// ParameterDeclProto is used to declare the expected type/shape
+// of a given attribute or input/output parameter.
+message ParameterDeclProto {
+  optional string name = 1;
+  optional TypeProto type = 2;
+  optional TensorShapeProto shape = 3;
+// An optional human-readable documentation for this parameter.
+  optional string doc_string = 4;
+};
+
+// SignatureDeclProto describes the expected attributes and input/output parameters
+// for a given operator.
+message SignatureDeclProto {
+// The formal input parameters to the operation or function 
+  repeated ParameterDeclProto input_params = 1;
+// The formal output parameters to the operation or function 
+  repeated ParameterDeclProto output_params = 2;
+// The declaration of expected attributes to the operation or function 
+  repeated ParameterDeclProto input_attributes = 3; 
```

Comment:
you might need to add a notion of optional / default attributed. Otherwise optional args might easily lead to combinatorial explosion of OperatorDecls (assuming that you intend to produce a separate one for each combination)